llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp

```diff
@@ -1841,12 +1842,74 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
   return false;
 }
 
+/// Memmove calls with overlapping src/dest buffers that come after a memset may
+/// be removed.
+bool MemCpyOptPass::isMemMoveMemSetDependency(MemMoveInst *M) {
+  const auto &DL = M->getDataLayout();
+  MemoryUseOrDef *MemMoveAccess = MSSA->getMemoryAccess(M);
+  if (!MemMoveAccess)
+    return false;
+
+  // The memmove is of form memmove(x, x + A, B).
+  MemoryLocation SourceLoc = MemoryLocation::getForSource(M);
+  auto *MemMoveSourceOp = M->getSource();
+  auto *Source = dyn_cast<GEPOperator>(MemMoveSourceOp);
+  if (!Source)
+    return false;
+
+  APInt Offset(DL.getIndexTypeSizeInBits(Source->getType()), 0);
+  LocationSize MemMoveLocSize = SourceLoc.Size;
+  if (Source->getPointerOperand() != M->getDest() ||
+      !MemMoveLocSize.hasValue() || Offset.isNegative() ||
+      !Source->accumulateConstantOffset(DL, Offset)) {
+    return false;
+  }
+
+  const uint64_t MemMoveSize = MemMoveLocSize.getValue();
+  LocationSize TotalSize =
+      LocationSize::precise(Offset.getZExtValue() + MemMoveSize);
+  MemoryLocation CombinedSourceLoc(MemMoveSourceOp, TotalSize);
+  MemoryLocation CombinedDestLoc(M->getDest(), TotalSize);
+
+  // The first dominating clobbering MemoryAccess for the combined location
+  // needs to be a memset.
+  BatchAAResults BAA(*AA);
+  MemSetInst *MS = nullptr;
+  MemoryAccess *FirstDef = MemMoveAccess->getDefiningAccess();
+  MemoryAccess *DestClobber = MSSA->getWalker()->getClobberingMemoryAccess(
+      FirstDef, CombinedDestLoc, BAA);
```

Comment:
```suggestion
  auto *DestClobber = dyn_cast<MemoryDef>(MSSA->getWalker()->getClobberingMemoryAccess(
      FirstDef, CombinedDestLoc, BAA));
  if (!DestClobber)
    return nullptr;
```
And then you can directly assign `MS` after that.